### PR TITLE
NIFI-3238 - Introduce ListenBeats processor and deprecates ListenLumbe…

### PR DESCRIPTION
**Pull request by trixpan**
…rjack processor

Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [x] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [x] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by trixpan** [comment]
@mattyb149 - this isn't ready yet but I suspect it may be functional. So far I have been testing with filebeat but it should work with metricbeat as well.

Of note:

- the processor performs partial ACKs, more precisely it replies to EVERY sequence number with an ACK, completely ignoring the window size. This may change prior to merging.

- Given ListenLumberjack has been my first attempt in coding, while migrating from ListenLumberjack to ListenBeats I tried to clean the code a bit, I think there still space for improvement.

To test the processor all it is needed is to:

1. Download filebeat and unzip it
2. use the following config:

```
filebeat.prospectors:
- input_type: log
  paths:
    - logsource/*.log

output.logstash:
  hosts: ["localhost:5044"]
  compression_level: 0
  timeout: 600
  bulk_max_size: 1

```

**Comment by scottcrespo** [comment]
Hey @trixpan 

@mattyb149 linked me to this PR since I'm an [Elastic Beats](https://github.com/elastic/beats) contributor and NiFi fan boy. 

I just started working on a [Metricbeat module](https://github.com/elastic/beats/pull/3374) to monitor NiFi via its REST API. I'd love to talk with you to better understand your work here, and discuss an implementation for Metricbeat. 

You can email me at: sccrespo@gmail.com
Event better, hit me up on skype: scott.crespo


Happy Coding!

**Comment by trixpan** [comment]
@scottcrespo thanks for the message. 

the work around here is fairly simple in nature: to allow any beat (filebeat, metricbeat, etc) to send their data directly into NiFi, allowing people currently using logstash to seamlessly migrate data pipelines into NiFi.

When you read any beats doco you will note they have a few different outputs: Kafka, ES and Logstash. This processor allows NiFi to act as a destination of _*beats_ logstash output.

So in theory, your Metricbeat module could still use metricbeat as underlying agent, but instead of shipping it directly into ES as I believe you are planning to do, ship it to NiFi (so it can be further processed, persisted in HDFS, etc)

**Comment by nickcarenza** [comment]
I've been playing with this branch locally for a few days and I haven't experienced any issues yet.

**Comment by trixpan** [comment]
@nickcarenza thanks for the feedback

**Comment by trixpan** [comment]
For future reference:

https://discuss.elastic.co/t/testers-wanted-beat-integration-with-apache-nifi/73671/

**Comment by mattyb149** [comment]
Reviewing...

File: nifi-nar-bundles/nifi-beats-bundle/nifi-beats-processors/src/main/java/org/apache/nifi/processors/beats/ListenBeats.java [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        final List<BeatsEvent> events = batch.getEvents();
+        // the sender and command will be the same for all events based on the batch key
+        final String sender = events.get(0).getSender();
+        final int numAttributes = events.size() == 1 ? 5 : 4;
+        final Map<String, String> attributes = new HashMap<>(numAttributes);
+        attributes.put(beatsAttributes.SENDER.key(), sender);
+        attributes.put(beatsAttributes.PORT.key(), String.valueOf(port));
+        attributes.put(CoreAttributes.MIME_TYPE.key(), "application/json");
+        // if there was only one event then we can pass on the transaction
+        // NOTE: we could pass on all the transaction ids joined together
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-beats-bundle/nifi-beats-processors/src/main/java/org/apache/nifi/processors/beats/ListenBeats.java]: maybe move this comment up to the def of numAttributes, to explain why there might be an extra one

File: nifi-nar-bundles/nifi-lumberjack-bundle/nifi-lumberjack-processors/src/main/java/org/apache/nifi/processors/lumberjack/ListenLumberjack.java [hunk 2]
```diff
@@ -62,12 +62,13 @@
 
 import com.google.gson.Gson;
 
+@Deprecated
 @InputRequirement(InputRequirement.Requirement.INPUT_FORBIDDEN)
 @Tags({"listen", "lumberjack", "tcp", "logs"})
-@CapabilityDescription("Listens for Lumberjack messages being sent to a given port over TCP. Each message will be " +
+@CapabilityDescription("This processor is deprecated and will be removed in the near future. Listens for Lumberjack messages being sent to a given port over TCP. Each message will be " +
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-lumberjack-bundle/nifi-lumberjack-processors/src/main/java/org/apache/nifi/processors/lumberjack/ListenLumberjack.java]: maybe rephrase to "may be removed from the distribution". Deprecated things have a tendency to linger ;)

File: nifi-nar-bundles/nifi-lumberjack-bundle/nifi-lumberjack-processors/src/main/java/org/apache/nifi/processors/lumberjack/event/LumberjackEvent.java [hunk 3]
```diff
@@ -24,6 +24,7 @@
 /**
  * A Lumberjack event which adds the transaction number and command to the StandardEvent.
  */
+@Deprecated
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-lumberjack-bundle/nifi-lumberjack-processors/src/main/java/org/apache/nifi/processors/lumberjack/event/LumberjackEvent.java]: For all @Deprecated tags, please provide @deprecated Javadoc for why it is being deprecated and what they should switch to

File: nifi-nar-bundles/nifi-beats-bundle/nifi-beats-processors/pom.xml [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.11</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.google.code.gson</groupId>
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-beats-bundle/nifi-beats-processors/pom.xml]: if this is a compile dependency and "you're already in there", please move this up above the "test" and "provided" dependencies. Not a requirement but a suggestion :)

File: nifi-nar-bundles/nifi-beats-bundle/nifi-beats-nar/src/main/resources/META-INF/LICENSE [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+   limitations under the License.  
+
+APACHE NIFI SUBCOMPONENTS:
+
+The Apache NiFi project contains subcomponents with separate copyright
+notices and license terms. Your use of the source code for the these
+subcomponents is subject to the terms and conditions of the following
+licenses. 
+
+  The binary distribution of this product bundles 'Bouncy Castle JDK 1.5'
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-beats-bundle/nifi-beats-nar/src/main/resources/META-INF/LICENSE]: Please double-check [NIFI-2954](https://issues.apache.org/jira/browse/NIFI-2954) with respect to what needs to go in the L&N as well as what can be specified as "provided" by the parent NAR(s) / framework. I couldn't immediately tell if all was well so I thought I'd point it out just to be sure.

File: nifi-nar-bundles/nifi-beats-bundle/nifi-beats-processors/src/main/java/org/apache/nifi/processors/beats/frame/BeatsDecoder.java [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+import org.slf4j.LoggerFactory;
+
+/**
+ * Decodes a Beats frame by maintaining a state based on each byte that has been processed. This class
+ * should not be shared by multiple threads.
+ */
+public class BeatsDecoder {
+
+
+    static final Logger logger = LoggerFactory.getLogger(BeatsDecoder.class);
```
Comment (trixpan) [inline nifi-nar-bundles/nifi-beats-bundle/nifi-beats-processors/src/main/java/org/apache/nifi/processors/beats/frame/BeatsDecoder.java]: sure. Will adjust.

File: nifi-nar-bundles/nifi-lumberjack-bundle/nifi-lumberjack-processors/src/main/java/org/apache/nifi/processors/lumberjack/ListenLumberjack.java (hunk 2 above)
Comment (trixpan) [inline nifi-nar-bundles/nifi-lumberjack-bundle/nifi-lumberjack-processors/src/main/java/org/apache/nifi/processors/lumberjack/ListenLumberjack.java]: will rephrase

File: nifi-nar-bundles/nifi-beats-bundle/nifi-beats-nar/src/main/resources/META-INF/LICENSE (hunk 5 above)
Comment (trixpan) [inline nifi-nar-bundles/nifi-beats-bundle/nifi-beats-nar/src/main/resources/META-INF/LICENSE]: I missed that JIRA. Will help with other of my PRs as well. :-)

File: nifi-nar-bundles/nifi-beats-bundle/nifi-beats-processors/src/main/java/org/apache/nifi/processors/beats/ListenBeats.java (hunk 1 above)
Comment (trixpan) [inline nifi-nar-bundles/nifi-beats-bundle/nifi-beats-processors/src/main/java/org/apache/nifi/processors/beats/ListenBeats.java]: fair enough.

File: nifi-nar-bundles/nifi-beats-bundle/nifi-beats-nar/src/main/resources/META-INF/LICENSE (hunk 5 above)
Comment (trixpan) [inline nifi-nar-bundles/nifi-beats-bundle/nifi-beats-nar/src/main/resources/META-INF/LICENSE]: Ok. Double checked, the LICENSE file follows the same line added by Joe as part of NIFI-2954, so I reckon it is ok as it is.

File: nifi-nar-bundles/nifi-beats-bundle/nifi-beats-processors/pom.xml (hunk 4 above)
Comment (trixpan) [inline nifi-nar-bundles/nifi-beats-bundle/nifi-beats-processors/pom.xml]: fixed

File: nifi-nar-bundles/nifi-lumberjack-bundle/nifi-lumberjack-processors/src/main/java/org/apache/nifi/processors/lumberjack/ListenLumberjack.java (hunk 2 above)
Comment (trixpan) [inline nifi-nar-bundles/nifi-lumberjack-bundle/nifi-lumberjack-processors/src/main/java/org/apache/nifi/processors/lumberjack/ListenLumberjack.java]: fixed

File: nifi-nar-bundles/nifi-lumberjack-bundle/nifi-lumberjack-processors/src/main/java/org/apache/nifi/processors/lumberjack/event/LumberjackEvent.java (hunk 3 above)
Comment (trixpan) [inline nifi-nar-bundles/nifi-lumberjack-bundle/nifi-lumberjack-processors/src/main/java/org/apache/nifi/processors/lumberjack/event/LumberjackEvent.java]: hopefully addressed

File: nifi-nar-bundles/nifi-beats-bundle/nifi-beats-processors/src/main/java/org/apache/nifi/processors/beats/ListenBeats.java (hunk 1 above)
Comment (trixpan) [inline nifi-nar-bundles/nifi-beats-bundle/nifi-beats-processors/src/main/java/org/apache/nifi/processors/beats/ListenBeats.java]: wait a second. it is already there:

@WritesAttribute(attribute = "beats.sequencenumber", description = "The sequence number of the message. Only included if <Batch Size> is 1."),

File: nifi-nar-bundles/nifi-beats-bundle/nifi-beats-processors/src/main/java/org/apache/nifi/processors/beats/frame/BeatsDecoder.java (hunk 6 above)
Comment (trixpan) [inline nifi-nar-bundles/nifi-beats-bundle/nifi-beats-processors/src/main/java/org/apache/nifi/processors/beats/frame/BeatsDecoder.java]: fixed

File: nifi-nar-bundles/nifi-beats-bundle/nifi-beats-processors/src/main/java/org/apache/nifi/processors/beats/frame/BeatsDecoder.java [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+    static final int COMPRESSED_MIN_LENGTH = MIN_FRAME_HEADER_LENGTH + 4; // 32 bit unsigned + payload
+    static final int JSON_MIN_LENGTH = MIN_FRAME_HEADER_LENGTH + 8; // 32 bit unsigned sequence number + 32 bit unsigned payload length
+
+    public static final byte FRAME_WINDOWSIZE = 0x57, FRAME_DATA = 0x44, FRAME_COMPRESSED = 0x43, FRAME_ACK = 0x41, FRAME_JSON = 0x4a;
+
+    /**
+     * @param charset the charset to decode bytes from the frame
+     */
+    public BeatsDecoder(final Charset charset) {
+        this(charset, new ByteArrayOutputStream(4096));
```
Comment (trixpan) [inline nifi-nar-bundles/nifi-beats-bundle/nifi-beats-processors/src/main/java/org/apache/nifi/processors/beats/frame/BeatsDecoder.java]: I suspect so. I tested with both smaller and larger values and they all work ok. Happy to convert into a property tho so it can be adjusted.

Comment (trixpan) [inline nifi-nar-bundles/nifi-beats-bundle/nifi-beats-processors/src/main/java/org/apache/nifi/processors/beats/frame/BeatsDecoder.java]: correction: passing as attribute will be painful. 4096 seems to be the value of choice around NiFi. Small values also work so I would reckon except for performance, no significant issue should be cause by the default

File: nifi-nar-bundles/nifi-beats-bundle/nifi-beats-processors/src/main/java/org/apache/nifi/processors/beats/ListenBeats.java [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+        session.commit();
+        // respond to each event to acknowledge successful receipt
+        for (final BeatsEvent event : events) {
+            respond(event, BeatsResponse.ok(event.getSeqNumber()));
+        }
+    }
+
+    @Override
+    protected String getTransitUri(FlowFileEventBatch batch) {
+        final String sender = batch.getEvents().get(0).getSender();
```
Comment (trixpan) [inline nifi-nar-bundles/nifi-beats-bundle/nifi-beats-processors/src/main/java/org/apache/nifi/processors/beats/ListenBeats.java]: I would say not:

final InetAddress sender = socketChannel.socket().getInetAddress();

But some people also assumed the Titanic was unsinkable... :ship:

**Comment by trixpan** [comment]
@mattyb149 - hopefully addressed your feedback.

File: nifi-nar-bundles/nifi-beats-bundle/nifi-beats-processors/src/main/java/org/apache/nifi/processors/beats/ListenBeats.java [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+import org.apache.nifi.processors.beats.event.BeatsEventFactory;
+import org.apache.nifi.processors.beats.frame.BeatsEncoder;
+import org.apache.nifi.processors.beats.handler.BeatsSocketChannelHandlerFactory;
+import org.apache.nifi.processors.beats.response.BeatsChannelResponse;
+import org.apache.nifi.processors.beats.response.BeatsResponse;
+import org.apache.nifi.ssl.SSLContextService;
+
+@InputRequirement(InputRequirement.Requirement.INPUT_FORBIDDEN)
+@Tags({"listen", "beats", "tcp", "logs"})
+@CapabilityDescription("Listens for messages sent by libbeat compatible clients (e.g. filebeats, metricbeats, etc) being sent to a given port over TCP, writing its contents in" +
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-beats-bundle/nifi-beats-processors/src/main/java/org/apache/nifi/processors/beats/ListenBeats.java]: Can you mention here (just for completeness) that the processor wants the "logstash" type output vs the Elasticsearch output? Might be a nod to those who have forgotten to configure their beats correctly.

**Comment by mattyb149** [comment]
+1 LGTM, ran tests and tried with a live NiFi using metricbeat and filebeat. Thanks very much for the great contribution!  Merging to master
